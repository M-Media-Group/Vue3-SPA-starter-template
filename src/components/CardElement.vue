<script setup lang="ts">
import type { PropType } from "vue";

defineProps({
  // The title of the card
  title: {
    type: String,
    required: false,
  },
  // The subtitle of the card
  subtitle: {
    type: String,
    required: false,
  },
  // The image of the card
  images: {
    // Images is an array of images, each image contains src and alt
    type: Array as PropType<{ src: string; alt: string }[]>,
    required: false,
  },
});
</script>

<template>
  <article>
    <div class="images" v-if="images">
      <img
        v-for="image in images"
        :src="image.src"
        :alt="image.alt"
        :key="image.alt"
      />
    </div>

    <header v-if="title || subtitle || $slots.headerActions || $slots.header">
      <slot name="header">
        <div>
          <h3 v-if="title">{{ title }}</h3>
          <p v-if="subtitle">{{ subtitle }}</p>
        </div>
        <div class="actions" v-if="$slots.headerActions">
          <slot name="headerActions" />
        </div>
      </slot>
    </header>
    <div>
      <slot />
    </div>
    <footer v-if="$slots.footer">
      <slot name="footer" />
    </footer>
  </article>
</template>
