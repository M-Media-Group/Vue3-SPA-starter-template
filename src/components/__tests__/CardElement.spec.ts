import { describe, expect, it } from "vitest";

import { mount } from "@vue/test-utils";
import CardElement from "../CardElement.vue";

describe("Card element", () => {
  it("renders properly without footer", () => {
    const wrapper = mount(CardElement, {
      props: {
        title: "Hello",
        subtitle: "subtitle",
        images: [
          {
            src: "https://picsum.photos/200/300",
            alt: "alt",
          },
        ],
      },
    });

    // Expect an article element to exist
    expect(wrapper.find("article").exists()).toBe(true);

    // Expect the texts to be present
    expect(wrapper.text()).toContain("Hello");
    expect(wrapper.text()).toContain("subtitle");

    // Expect an image to exist
    expect(wrapper.find("img").exists()).toBe(true);

    // Expect the image to have the correct src
    expect(wrapper.find("img").attributes("src")).toBe(
      "https://picsum.photos/200/300"
    );

    // There should be a header element in the article
    expect(wrapper.find("article").find("header").exists()).toBe(true);

    // There should be no footer since we didn't pass it into the slot
    expect(wrapper.find("footer").exists()).toBe(false);
  });

  it("shows a footer when one is passed", () => {
    const wrapper = mount(CardElement, {
      props: {
        title: "Hello",
        subtitle: "subtitle",
        images: [
          {
            src: "https://picsum.photos/200/300",
            alt: "alt",
          },
        ],
      },
      slots: {
        footer: "Footer",
      },
    });

    // There should be a footer since we pass it into the slot
    expect(wrapper.find("footer").exists()).toBe(true);

    // The footer should contain the text "Footer"
    expect(wrapper.find("footer").text()).toBe("Footer");
  });

  it("shows a card title when one is passed", () => {
    const wrapper = mount(CardElement, {
      props: {
        title: "Hello",
      },
    });

    // There should be a title since we pass it into the props. Find the text
    // inside the header element
    expect(wrapper.find("header").text()).toContain("Hello");
  });

  it("shows a card subtitle when one is passed", () => {
    const wrapper = mount(CardElement, {
      props: {
        title: "Hello",
        subtitle: "subtitle",
      },
    });

    // There should be a title since we pass it into the props. Find the text
    // inside the header element
    expect(wrapper.find("header").text()).toContain("subtitle");
  });

  it("shows images when passed", () => {
    const wrapper = mount(CardElement, {
      props: {
        title: "Hello",
        subtitle: "subtitle",
        images: [
          {
            src: "https://picsum.photos/200/300",
            alt: "alt",
          },
        ],
      },
    });

    // There should be an image since we pass it into the props. Find the image
    // inside the wrapper
    expect(wrapper.find("img").exists()).toBe(true);

    // The image should have the correct src
    expect(wrapper.find("img").attributes("src")).toBe(
      "https://picsum.photos/200/300"
    );
  });

  it("shows body content when passed to default slot", () => {
    const wrapper = mount(CardElement, {
      props: {
        title: "Hello",
        subtitle: "subtitle",
      },
      slots: {
        default: "Body content",
      },
    });

    // There should be a body since we pass it into the
    // default slot. Find the text inside the wrapper
    expect(wrapper.text()).toContain("Body content");
  });

  it("shows headerActions when that slot is passed", () => {
    const wrapper = mount(CardElement, {
      props: {
        title: "Hello",
        subtitle: "subtitle",
      },
      slots: {
        headerActions: "Header actions",
      },
    });

    // There should be a headerActions since we pass it into the
    // headerActions slot. Find the text inside the wrapper
    expect(wrapper.text()).toContain("Header actions");
  });
});
