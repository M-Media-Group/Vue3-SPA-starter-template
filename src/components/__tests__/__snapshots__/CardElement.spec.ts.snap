// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card element > renders properly without footer 1`] = `
"<!-- @todo the animation breaks with component, need to fix. It works when article is direct child or when <template> (from Vue) is used, but not with <component>, even with an :is to a Vue template -->
<vue:template>
  <article>
    <div class="images overflow-auto" tabindex="0"><img src="https://picsum.photos/200/300" alt="alt"></div>
    <header>
      <hgroup>
        <h3>Hello</h3>
        <p>subtitle</p>
      </hgroup>
      <!--v-if-->
    </header>
    <!--v-if-->
  </article>
</vue:template>"
`;
